visual_behavior/data_access/utilities.py

```diff
@@ -1268,6 +1268,96 @@ def correct_dataframe_filepath(dataframe, column_string):
+    for each mouse_id from the ophys_experiments_table and merge that into the behavior_sessions table so that we can
+    filter the behavior_sessions by project_code
+    """
+    # get project codes for each mouse from experiments table
+    mouse_projects = ophys_experiments_table.drop_duplicates(subset=['mouse_id', 'project_code'])
+    mouse_projects = mouse_projects[['mouse_id', 'project_code']]
+
+    # merge into behavior sessions
+    print(len(behavior_sessions_table), 'behavior sessions in original behavior_sessions table')
+    behavior_sessions = behavior_sessions_table.drop(columns='project_code').merge(mouse_projects, on='mouse_id',
+                                                                                   how='left')
+    print(len(behavior_sessions), 'behavior sessions after merging with project codes')
+
+    return behavior_sessions
+
+
+def add_training_stage_info_to_behavior_sessions(behavior_sessions):
+    """
+    Adds a column to behavior_sessions called `training_stage` that is the first two elements of `session_type`,
+    i.e. 'TRAINING_2', or 'OPHYS_3',
+    as well as Boolean columns indicating whether a given session was the first or last day of each `training_stage`
+    """
+    # create training_stage column as abbreviation of session_type
+    behavior_sessions['training_stage'] = [stage.split('_')[0] + '_' + stage.split('_')[1] for stage in
+                                           behavior_sessions.session_type.values]
+
+    # add first day of stage based on acquisition date
+    behavior_sessions['first_day_of_stage'] = False
+    stage_start = behavior_sessions.sort_values(by=['mouse_id', 'date_of_acquisition'])
+    stage_start = stage_start.drop_duplicates(subset=['mouse_id', 'training_stage'])
+    behavior_sessions.at[stage_start.index.values, 'first_day_of_stage'] = True
+
+    # add last day of stage based on acquisition date
+    behavior_sessions['last_day_of_stage'] = False
+    stage_end = behavior_sessions.sort_values(by=['mouse_id', 'date_of_acquisition'], ascending=False)
+    stage_end = stage_end.drop_duplicates(subset=['mouse_id', 'training_stage'])
+    behavior_sessions.at[stage_end.index.values, 'last_day_of_stage'] = True
+
+    return behavior_sessions
+
+
+def add_experience_level_to_behavior_sessions(behavior_sessions):
+    """
+    adds a column to behavior_sessions table that contains a string indicating whether a session had
+    exposure level of Familiar, Novel 1, or Novel >1, based on session number and prior_exposure_to_image_set for ophys sessions,
+    then assign all TRAINING sessions with images as 'Familiar',
+    and any TRAINING sessions with gratings as 'Gratings'
+
+    input df must have 'session_number' column which can be added / update using the add_session_number_to_experiments_table function
+    """
+    # add experience_level column with strings indicating relevant conditions
+    behavior_sessions['experience_level'] = 'None'
```

Comment:
Should this either be `None` (as python type, not a string), or `"Training"`? Having a string with "None" is confusing